### DISCUSSION TITLE
Reason for selectively using create() method in TinyALU example?

### DISCUSSION THREAD

**Discussion by denchacko** [pyuvm/pyuvm#137] [Q&A] [answered]
Hi,

In the TinyALU example, I noticed that the create() method has only been used selectively for the Driver and TestAllSeq instances and hasn't been used anywhere else.
Is there any particular reason for doing it this way?
Ideally, we should use the create() method for all uvm_objects/uvm_components to allow scope for factory overriding, right?

Regards,
Dennis

**Comment by raysalemi** [comment, marked as answer]
Yes, I have a reason I do it this way.

I believe the component developer should control which components can be overridden.  Otherwise, users could override the wrong things, break the component, and blame the developer.

Every overridable component needs to be documented so it can be overridden properly.  So, we either document the component or use regular instantiation.

**Reply by denchacko** [reply]
I see.
Would the same reasoning apply to traditional SV-based UVM? I believe that generally the create() method approach is used for almost all components.

**Reply by raysalemi** [reply]
> I see.
> 
> Would the same reasoning apply to traditional SV-based UVM? I believe that generally the create() method approach is used for almost all components.
> 
> 

Yes.  I have always done it this way (SV or Python) for the same reason.

This is, of course, a matter of style.
